app.py

```diff
@@ -1151,16 +1188,24 @@ def list_programs(user):
     return {'programs': DATABASE.programs_for_user(user['username'])}
 
 
-# Not very restful to use a GET to delete something, but indeed convenient; we can do it with a single link and avoiding AJAX.
-@app.route('/programs/delete/<program_id>', methods=['GET'])
+@app.route('/programs/delete/', methods=['DELETE'])
```

Comment:
I agree with you that the `GET` here was wrong, and this needs changing, I don't think `DELETE` goes with this URL structure. `DELETE` should not have a request body: the URL alone should identify the resource to be deleted.

Any of the following would make sense:

HTTP verb | Parameters | Interpretation
-----------|------------|----------------
DELETE /programs/1234 | |  `/programs/1234` is a resource which we're asking to DELETE
POST /programs |  { command: 'delete', id: 1324 } |  send a 'delete' command to an endpoint that manages programs
POST /programs/delete | { id: 1234 } | send an 'id' to an endpoint that only does deletes
POST /programs/1234/delete | | invoke the 'delete' command of program 1234
POST /programs/delete/1234 | | alternative URL structure for the previous one
```

Of those, `DELETE /programs/1234` is the most HTTP-pure. However, one might ask: what do we buy for that? The API is not for other people to consume, only for us, and we might as well do what is convenient.

I myself used to have strong opinions on how REST APIs are "better" than RPC APIs (just like I had very strong opinions on brace styles)... but then I worked on some projects and found it didn't actually matter all that much and I didn't care anymore about style or purity, only about what was (a) correct and (b) convenient.

Long story short: I agree that `GET` is wrong and should be changed. Either POST or DELETE would be correct. The simplest change that will make our broken system correct is by changing GET to POST, and leaving everything else the same, in my opinion.

Of course, free advice is worth exactly what you're paying for it. Feel free to do with this opinion what you want.


(Another two cents: seems like this PR is growing quite large, and this change in particular doesn't seem like it has anything to do with the title of the PR. Are you trying to do too many things at once? You might want to make different PRs out of different logical changes to make it easier to review, discuss and approve them in isolation).

Reply:
Thanks for the advice! I will re-write it to be conform one of your options. This PR is indeed getting a bit too big for its purpose. This is mainly due to having to re-write some stuff to enable us to return a response with possible achievements. (As re-writing the GET into a DELETE). So all is still related to the achievements, but got a big out of hand while implementing them.

Reply:
> This is mainly due to having to re-write some stuff to enable us to return a response with possible achievements.

I understand. But assuming those other changes are useful on their own (which it looks like they are), it is not going to be wasted effort to extract them, make individual PRs out of them, merge those first, and then continue with this big work.